File: compiler/typecheck/solver.go
package typecheck

import (
	"reflect"
	"slices"

	"wipple/database"
)

type Solver struct {
	Db               *database.Db
	Constraints      Constraints
	ImpliedInstances []Instance
	Progress         bool
	Error            bool
	groups           *groups
}

func NewSolver(db *database.Db) *Solver {
	return &Solver{
		Db:     db,
		groups: newGroups(nil),
	}
}

func CopySolver(other *Solver) *Solver {
	solver := NewSolver(other.Db)
	solver.Inherit(other)
	return solver
}

func (s *Solver) Inherit(other *Solver) {
	s.groups = newGroups(other.groups)
}

func (s *Solver) AppendGroup(group *Group) {
	s.groups.append(group, func(left, right *ConstructedType) bool {
		return s.unifyConstructedTypes(left, right)
	})
}

func (s *Solver) Run() bool {
	progress := false
	for {
		s.Progress = false
		s.RunPassUntil(nil)
		if !s.Progress {
			break
		}

		progress = true
	}

	// Run a final pass
	s.RunPassUntil(nil)

	return progress || s.Progress
}

func (s *Solver) RunPassUntil(stop reflect.Type) {
	s.Constraints.RunUntil(s, stop)

	if !s.Progress {
		s.Constraints.RunDefaults(s)
	}
}

func (s *Solver) Imply(instance Instance) {
	if !slices.ContainsFunc(s.ImpliedInstances, func(existing Instance) bool {
		return existing.Node == instance.Node
	}) {
		s.ImpliedInstances = append(s.ImpliedInstances, instance)
	}
}

func (s *Solver) Unify(trace Constraint, left Type, right Type) {
	if left == right {
		return
	}

	var originalLeftNode database.Node
	originalLeftNode, leftWasNode := left.(database.Node)

	var originalRightNode database.Node
	originalRightNode, rightWasNode := right.(database.Node)

	if leftWasNode && rightWasNode {
		s.merge(trace, originalLeftNode, originalRightNode)
		s.Progress = true
		return
	}

	left = s.applyShallow(left)
	right = s.applyShallow(right)

	leftNode, leftIsNode := left.(database.Node)
	rightNode, rightIsNode := right.(database.Node)

	if leftIsNode && rightIsNode {
		s.merge(trace, leftNode, rightNode)
		s.Progress = true
	} else if leftIsNode {
		s.insert(trace, leftNode, right)
		s.Progress = true
	} else if rightIsNode {
		s.insert(trace, rightNode, left)
		s.Progress = true
	} else {
		left := left.(*ConstructedType)
		right := right.(*ConstructedType)

		ok := s.unifyConstructedTypes(left, right, originalLeftNode, originalRightNode)
		if !ok {
			s.Error = true

			// Report conflicts on the original nodes
			if leftWasNode {
				s.insert(trace, originalLeftNode, right)
			}
			if rightWasNode {
				s.insert(trace, originalRightNode, left)
			}
		}
	}
}

func (s *Solver) unifyConstructedTypes(left *ConstructedType, right *ConstructedType, originalNodes ...database.Node) bool {
	// Type parameters are unique
	if left.Instantiate != right.Instantiate {
		s.Error = true
		return false
	}

	if left.Tag == right.Tag {
		for i := 0; i < len(left.Children) && i < len(right.Children); i++ {
			leftChild := left.Children[i]
			rightChild := right.Children[i]

			if TypeReferencesNode(leftChild, originalNodes...) || TypeReferencesNode(rightChild, originalNodes...) {
				// Recursive types
				continue
			}

			s.Unify(nil, leftChild, rightChild)
		}
	}

	ok := left.Tag == right.Tag && len(left.Children) == len(right.Children)

	if !ok {
		s.Error = true
	}

	return ok
}

func (s *Solver) Apply(ty Type) Type {
	return TraverseType(ty, func(ty Type) (Type, bool) {
		return s.applyShallow(ty), false
	})
}

func (s *Solver) applyShallow(ty Type) Type {
	if node, ok := ty.(database.Node); ok {
		_, group, ok := s.groups.FindGroup(node)
		if !ok || len(group.Types) == 0 {
			return ty
		}

		return group.Types[0]
	}

	return ty
}

func (s *Solver) insert(trace Constraint, node database.Node, types ...Type) {
	if entry, group, ok := s.groups.FindGroup(node); ok {
		for _, ty := range types {
			switch ty := ty.(type) {
			case database.Node:
				s.merge(trace, node, ty)
				entry, group, _ = s.groups.FindGroup(node)
			case *ConstructedType:
				group = entry.Clone(group)
				group.Types = append(group.Types, ty)

				if trace != nil {
					group.Trace = append(group.Trace, trace)
				}
			}
		}

		group.normalize()

		return
	}

	groupNodes := []database.Node{node}
	groupTypes := make([]*ConstructedType, 0, 1)
	for _, ty := range types {
		switch ty := ty.(type) {
		case database.Node:
			groupNodes = append(groupNodes, ty)
		case *ConstructedType:
			groupTypes = append(groupTypes, ty)
		}
	}

	group := makeGroup(groupNodes, groupTypes, []Constraint{trace})
	group.normalize()

	s.AppendGroup(group)
}

func (s *Solver) merge(trace Constraint, leftNode database.Node, rightNode database.Node) {
	s.groups.merge(trace, leftNode, rightNode, func(left, right *ConstructedType) bool {
		return s.unifyConstructedTypes(left, right, leftNode, rightNode)
	})
}

func (s *Solver) Groups(order func(node database.Node) int) []*Group {
	var groups []*Group
	s.groups.Each(func(group *Group) {
		for i, ty := range group.Types {
			group.Types[i] = s.Apply(ty).(*ConstructedType)
		}

		group.normalize()

		slices.SortStableFunc(group.Nodes, func(left database.Node, right database.Node) int {
			return order(left) - order(right)
		})

		groups = append(groups, group)
	})

	return groups
}

File: compiler/src/typecheck/Solver.swift
import OrderedCollections

public struct GroupedWith: Fact { public var nodes: OrderedSet<Node> = [] }

class Solver {
    static let iterationLimit = 32

    var db: DB
    var groups = Groups()
    var impliedInstances: [Instance] = []
    var substitutionsToApply: [Substitutions] = []
    var progress = false
    var error = false
    var constraints = Constraints()
    private var iterations = 0

    init(db: DB) { self.db = db }

    func sortedGroups<Key: Comparable>(by key: (Node) -> Key) -> some Sequence<Group> {
        self.applyAll()
        return self.groups.sorted(by: key)
    }

    func run() {
        while true {
            self.progress = false

            self.runPass(until: nil)

            guard self.progress else { break }
        }

        // Run a final pass
        self.runPass(until: nil)

        // Apply recorded substitutions
        for substitutions in self.substitutionsToApply { self.apply(substitutions) }
        self.substitutionsToApply = []
    }

    func runPass(until stop: Constraint.Type?) {
        guard self.iterations < Self.iterationLimit else { return }

        self.constraints.run(with: self, until: stop)

        if !self.progress { self.constraints.runDefaults(with: self) }

        self.iterations += 1
    }

    func imply(_ instance: Instance) {
        guard !self.impliedInstances.lazy.map(\.node).contains(instance.node) else { return }

        self.impliedInstances.append(instance)
    }

    func unify(_ left: Node, _ right: Node, trace: Constraint?) {
        self.unify(.node(left), .node(right), trace: trace)
    }

    func unify(_ node: Node, _ type: ConstructedType, trace: Constraint?) {
        self.unify(.node(node), .constructed(type), trace: trace)
    }

    private func unify(_ left: Type, _ right: Type, trace: Constraint?) {
        if left == right { return }

        let originalLeftNode = left.node
        let originalRightNode = right.node

        if let originalLeftNode, let originalRightNode {
            self.progress = true
            self.merge(originalLeftNode, originalRightNode, trace: trace)
            return
        }

        let left = self.applyShallow(left)
        let right = self.applyShallow(right)

        switch (left, right) {
        case (.node(let left), .node(let right)):
            self.merge(left, right, trace: trace)
            self.progress = true
        case (.node(let node), .constructed(let type)), (.constructed(let type), .node(let node)):
            self.insert(node, type, trace: trace)
            self.progress = true
        case (.constructed(let left), .constructed(let right)):
            let originalNodes = OrderedSet(
                [originalLeftNode, originalRightNode].compactMap(\.self)
            )

            if !self.unify(left, right, originalNodes: originalNodes) {
                // Report conflicts on the original nodes
                if let originalLeftNode { self.insert(originalLeftNode, right, trace: trace) }
                if let originalRightNode { self.insert(originalRightNode, left, trace: trace) }
            }
        }
    }

    private func unify(
        _ left: ConstructedType,
        _ right: ConstructedType,
        originalNodes: OrderedSet<Node>,
    ) -> Bool {
        if left.tag == right.tag {
            for (leftChild, rightChild) in zip(left.children, right.children) {
                if !leftChild.referencedNodes.intersection(originalNodes).isEmpty
                    || !rightChild.referencedNodes.intersection(originalNodes).isEmpty
                {
                    // recursive types
                    continue
                }

                self.unify(leftChild, rightChild, trace: nil)
            }
        }

        if left.tag != right.tag || left.children.count != right.children.count {
            self.error = true
            return false
        }

        return true
    }

    func unify(_ left: Substitutions, _ right: Substitutions, trace: Constraint?) {
        for (parameter, left) in left {
            if let right = right[parameter] { self.unify(left, right, trace: trace) }
        }
    }

    private func merge(_ leftNode: Node, _ rightNode: Node, trace: Constraint?) {
        self.db[leftNode, GroupedWith.self, default: .init()].nodes.append(rightNode)
        self.db[rightNode, GroupedWith.self, default: .init()].nodes.append(leftNode)

        let leftIndex = self.groups.index(of: leftNode)
        let rightIndex = self.groups.index(of: rightNode)

        if (leftIndex != nil || rightIndex != nil) && leftIndex == rightIndex {
            return  // already the same group
        }

        let (index, group): (Int?, Group) =
            switch (leftIndex, rightIndex) {
            case (let leftIndex?, let rightIndex?):
                (leftIndex, self.groups.removeExisting(at: rightIndex))
            case (let leftIndex?, nil): (leftIndex, .init(nodes: [rightNode]))
            case (nil, let rightIndex?): (rightIndex, .init(nodes: [leftNode]))
            case (nil, nil): (nil, .init(nodes: [leftNode, rightNode]))
            }

        if let index {
            var newGroup = self.groups.removeExisting(at: index)

            self.groups.merge(group, into: &newGroup, trace: trace) { left, right in
                self.unify(left, right, originalNodes: [leftNode, rightNode])
            }

            _ = self.groups.insert(newGroup)
        } else {
            _ = self.groups.insert(group)
        }
    }

    private func insert(_ node: Node, _ type: ConstructedType, trace: Constraint?) {
        var type = type
        if type.representative == nil { type.representative = node }

        let insertInto = { (group: inout Group) in
            group.types.append(type)
            if let trace { group.trace.append(trace) }
        }

        if let index = self.groups.index(of: node) {
            insertInto(&self.groups[index])
        } else {
            var group = Group(nodes: [node])
            insertInto(&group)
            _ = self.groups.insert(group)
        }
    }

    func apply(_ type: Type) -> Type { type.traverse(with: self.applyShallow(_:)) }

    func apply(_ substitutions: Substitutions) {
        for (key, type) in substitutions { substitutions[key] = self.apply(type) }
    }

    private func applyShallow(_ type: Type) -> Type {
        guard case .node(let node) = type, let index = self.groups.index(of: node) else {
            return type
        }

        let group = self.groups[index]

        guard let type = group.types.first else { return type }

        return .constructed(type)
    }

    private func applyAll() {
        for index in self.groups.indices {
            var group = self.groups.removeExisting(at: index)

            for (index, type) in group.types.enumerated() {
                group.types[index] =
                    switch self.apply(.constructed(type)) {
                    case .constructed(let type): type
                    case .node(_): fatalError("constructed types remain constructed after `apply`")
                    }
            }

            self.groups.insertExisting(group, at: index)
        }
    }
}
